lib/oauth2/model/authorization.rb

```diff
@@ -0,0 +1,163 @@
+module OAuth2
+  module Model
+
+    class Authorization < ActiveRecord::Base
+      self.table_name = :oauth2_authorizations
+
+      belongs_to :oauth2_resource_owner, :polymorphic => true
+      alias :owner  :oauth2_resource_owner
+      alias :owner= :oauth2_resource_owner=
+
+      belongs_to :client, :class_name => 'OAuth2::Model::Client'
+
+      validates_presence_of :client, :owner
+
+      validates_uniqueness_of :code,               :scope => :client_id, :allow_nil => true
+      validates_uniqueness_of :refresh_token_hash, :scope => :client_id, :allow_nil => true
+      validates_uniqueness_of :access_token_hash,                        :allow_nil => true
+
+      attr_accessible nil if (defined?(ActiveRecord::VERSION) && ActiveRecord::VERSION::MAJOR <= 3) || defined?(ProtectedAttributes)
+
+      class << self
+        private :create, :new
+      end
+
+      extend Hashing
+      hashes_attributes :access_token, :refresh_token
+
+      def self.create_code(client: , additional_attributes: {}, predicate: nil)
+        Lib::SecureCodeScheme.generate( attributes: code_gen_attributes(additional_attributes),
+                                            predicate: predicate)
+      end
+
+      # When PKCE Atttributes (https://tools.ietf.org/html/rfc7636) are provided
+      # the code generation will use them as apposed to an opaque code
+      def self.code_gen_attributes(attr)
+        if attr.keys.include? CODE_CHALLENGE
+          {code_type: PKCE, code_challenge: attr[CODE_CHALLENGE], code_challenge_method: attr[CODE_CHALLENGE_METHOD]}
+        else
+          {code_type: OPAQUE}
+        end
+      end
+
+      def self.create_access_token
+        Lib::SecureCodeScheme.new.generate(predicate: ->(token) {
+          hash = Lib::SecureCodeScheme.new.hashify(token)
+          Helpers.count(self, :access_token_hash => hash).zero?
```

Comment:
This is original code logic. Looks like refresh_tokens are scoped by client for some reason.